kafka-connector/README.md

```diff
@@ -71,28 +61,72 @@ The resulting jar is at target/cps-kafka-connector.jar.
-| cps.project | String | REQUIRED (No default) | The project in Cloud Pub/Sub containing the topic to publish to. |
+| cps.topic | String | REQUIRED (No default) | The topic in Cloud Pub/Sub to publish to, e.g. "foo" for topic "/projects/bar/topics/foo". |
+| cps.project | String | REQUIRED (No default) | The project in Cloud Pub/Sub containing the topic, e.g. "bar" from above. |
 | maxBufferSize | Integer | 100 | The maximum number of messages that can be received for the messages on a topic partition before publishing them to Cloud Pub/Sub. |
+
+#### Schema Support and Data Model
+
+A pubsub message has two main parts: the message body and attributes. The
+message body is a ByteString object that translates well to and from the byte[]
+bodies of Kafka messages. For this reason, we recommend using primitive data
+types where possible to prevent deserializing and reserializing the same message
+body.
+
+Additionally, a Pubsub message size cannot exceed 10MB, so please check
+your broker's message.max.bytes configuration to prevent possible errors.
+
+The sink connector handles the conversion in the following way:
+
+*   For all primitive data types (integers, floats, bytes, and strings), the
+    bytes are stored as the message body.
+*   For map and struct types, the values are stored in attributes. Pubsub only
+    supports string to string mapping in attributes. To make the connector as
+    versatile as possible, the toString() method will be called on whatever
+    object passed in as the key or value for a map and the value for a struct.
+    *   One additional feature is we allow a specification of a particular
+        field or key to be placed in the Pubsub message body. To do so, set the
+        CPS_MESSAGE_BODY_NAME configuration with the struct field or map key.
+        This value is stored as a ByteString, any integer, byte, float, or array
+        type will be included in the message body as if it were the sole value.
+*   For arrays, we only support primitive array types due to potential
+    collisions of field names or keys of a struct/map array. The connector
+    handles arrays in a fairly predictable fashion, each value is concatenated
+    together into a ByteString object.
+*   In all cases, the Kafka key value is stored in the Pubsub message's
+    attributes as a string with the key set by
+    ConnectorUtils.CPS_MESSAGE_KEY_ATTRIBUTE, currently "key".
+
+The source connector takes a similar approach in handling the conversion
+from a Pubsub message into a SourceRecord with a relevant Schema.
+
+*   The connector searches for the given KAFKA_MESSAGE_KEY_CONFIG in the
+    attributes of the Pubsub message. If found, this will be used as the Kafka
+    key with a string schema type. Otherwise, it will be set to null.
+*   If the Pubsub message doesn't have any other attributes, the message body
+    is used as the SourceRecord value with a bytes schema.
```

Comment:
I don't think we want to talk in terms of SourceRecord. That is a concept that is purely a part of the Connector infrastructure and not inherent to Kafka as a whole. I don't think a user of the CPS connector would necessarily be interested in this fact. What I would say is that if one has messages attributes, then we recommend that they use a Converter that can structure data, e.g., a JSONConverter.